dsgrid/utils/spark.py

```diff
@@ -126,6 +129,10 @@ def restart_spark(*args, force=False, **kwargs):
         spark.stop()
         logger.info("Stopped the SparkSession so that it can be restarted with a new config.")
         spark = init_spark(*args, **kwargs)
+        if spark.conf.get("spark.sql.session.timeZone") != new_time_zone:
+            # We set this value in query_submitter.py and that change will get lost
+            # when the session is restarted.
+            spark.conf.set("spark.sql.session.timeZone", new_time_zone)
```

Comment:
`custom_spark_config` allows you to run code with a custom configuration and then revert the configuration. This is perfect for setting time zone. The spark session is not restarted.
`restart_spark_with_custom_conf` stops and restarts the spark session and optionally applies a custom configuration. I documented a reason [here](https://github.com/dsgrid/dsgrid/blob/main/dsgrid/project.py#L308). I was very puzzled at the time. I have found that we haven't needed to enable dynamic allocation in some time and restarting _may_ no longer be necessary.

The key point is that if you stop and restart the Spark session, all changes applied dynamically, like time zone, are lost.

Reply:
Is this code at the end only needed because Spark tries really hard to set the `spark.sql.session.timeZone` to your local time zone when you call `init_spark`? (I'm trying to understand why the `spark.sql.session.timeZone` configuration is special here. E.g., you check that this setting in `conf` is set properly, but none of the others.)

Reply:
Here is the order of increasing priority as I understand it:
- JVM, before Python starts
   - SPARK_CONF/spark-defaults.conf
   - parameters passed to pyspark or spark-submit
- After Python starts
   - parameters passed in Python to `SparkSession.builder.config(conf=conf).getOrCreate()`
   - A subset of parameters can be set dynamically afterwards, including time zone.

If you ever stop the spark session and re-create it, you get the settings at the time the first session was created (not just initial JVM settings). We set `spark.sql.parquet.outputTimestampType` in the builder and that sticks. The only setting that we are modifying afterwards is the time zone (new to this PR).

Reply:
Summary of check-in meeting conversation: We could generalize the above to be a list of non-sticky parameters we care about, but Dan does not expect that we'll need this for anything else.